### PR TITLE
Path dependent contracts

### PR DESCRIPTION
**Pull request by vargala4**

### CODE REVIEW

File: src/contract.py [hunk 1]
```diff
@@ -1,21 +1,34 @@
...
+from statistics import mean
 
 
 class Contract(ABC):
-    def __init__(self, und: Stock, dtype: PutCallFwd, strk: float, exp: float) -> None:
+    def __init__(self, und: Stock, dtype: PutCallFwd, longshort: LongShort, strk: float, exp: float, num_mon: int = 1) -> None:
         self._underlying: Stock = und
         self._derivative_type: PutCallFwd = dtype
+        self._derivative_longshort: LongShort = longshort
+        self.ls = 1 if longshort == LongShort.LONG else -1
```
Comment (friedgab) [inline src/contract.py:15]: Let's have this as `_ls` to indicate that no one should modify it outside of the class. It's a non-public property.

File: src/contract.py [hunk 2]
```diff
@@ -63,74 +87,322 @@ def get_timeline(self) -> list[float]:
...
+    def get_inout(self) -> InOut:
+        return self._inout
+
+    def is_breached(self, prices_und) -> bool:
+        if self._updown == 'UP':
+            return( any([self._barrier <= price for price in prices_und]) )
+        else:
+            return (any([self._barrier >= price for price in prices_und]))
+
+    def to_dict(self) -> dict[str, any]:
```
Comment (friedgab) [inline src/contract.py:295]: No need to re-define the whole dictionary. Hint:
```
super().to_dict().update({...})
```

File: src/contract.py [hunk 3]
```diff
@@ -63,74 +87,322 @@ def get_timeline(self) -> list[float]:
...
+
+    def get_updown(self) -> UpDown:
+        return self._updown
+
+    def get_inout(self) -> InOut:
+        return self._inout
+
+    def is_breached(self, prices_und) -> bool:
+        if self._updown == 'UP':
+            return( any([self._barrier <= price for price in prices_und]) )
```
Comment (friedgab) [inline src/contract.py:291]: Unnecessary parentheses around `any`.

File: src/contract.py [hunk 4]
```diff
@@ -63,74 +87,322 @@ def get_timeline(self) -> list[float]:
...
+        else:
+            self._raise_incorrect_derivative_type()
+
+
+class GenericContract(PathDependentContract):
+
+    def __init__(self, contract: ContractType, und: Stock, dtype: PutCallFwd, longshort: LongShort, strk: float, exp: float,
+                 barrier: float = np.Inf, updown: UpDown = None, inout: InOut = None) -> None:
+        if dtype not in [PutCallFwd.CALL, PutCallFwd.PUT, PutCallFwd.FWD]:
+            self._raise_incorrect_derivative_type()
```
Comment (friedgab) [inline src/contract.py:266]: Let's override this function in GenericContract as the error message would not be correct.

File: src/contract.py [hunk 5]
```diff
@@ -45,6 +63,12 @@ def get_timeline(self) -> list[float]:
     def _raise_incorrect_derivative_type(self):
         raise TypeError(f'Derivative type of {type(self).__name__} must be CALL or PUT')
 
+    def _raise_incorrect_barrier_updown_type(self):
```
Comment (friedgab) [inline src/contract.py:66]: These barrier related functions can be moved to `PathDependentContract`.

File: src/contract.py [hunk 6]
```diff
@@ -63,74 +87,322 @@ def get_timeline(self) -> list[float]:
...
+            "longshort": self._derivative_longshort,
+            "strike": self._strike,
+            "expiry": self._expiry,
+            "barrier": self._barrier,
+            "updown": self._updown,
+            "inout": self._inout
+        }
+
+    def convert_to_generic(self) -> GenericContract:
+        pass
```
Comment (friedgab) [inline src/contract.py:309]: `return self`

File: src/contract.py [hunk 7]
```diff
@@ -63,74 +87,322 @@ def get_timeline(self) -> list[float]:
...
             self._raise_incorrect_derivative_type()
 
 
-class GenericContract:
-    # todo: needs to be implemented, defined temporarily as empty class due to annotations
-    pass
+class EuropeanBarrierContract(PathDependentContract):
+
+    def __init__(self, und: Stock, dtype: PutCallFwd, longshort: LongShort, strk: float, exp: float,
+                 barrier: float, updown: UpDown, inout: InOut) -> None:
```
Comment (friedgab) [inline src/contract.py:203]: I would create a Barrier class for all barrier related property and functionality. Fine as it is for now, we can refactor later.

File: src/contract.py [hunk 8]
```diff
@@ -63,74 +87,322 @@ def get_timeline(self) -> list[float]:
...
+            return self.ls * float(spot - self._strike > 0)
         elif self._derivative_type == PutCallFwd.PUT:
-            return max(self._strike - spot, 0)
+            return self.ls * float(self._strike - spot > 0)
         else:
             self._raise_incorrect_derivative_type()
 
 
-class EuropeanDigitalContract(VanillaOptionContract):
+class PathDependentContract(Contract):
```
[the remark this answers is not in the document]
Comment (vargala4) [reply src/contract.py:162]: Done

File: src/enums.py [hunk 9]
```diff
@@ -1,5 +1,13 @@
 from enum import Enum
 
+class ContractType(str, Enum):
```
[the remark this answers is not in the document]
Comment (vargala4) [reply src/enums.py:3]: With this method we get a string for the type of the contracts, while with the original implementation ContractType is of Enum type. I thought we want to define every type of fixed name as an Enum (Stock, PutCall, etc).

File: src/contract.py [hunk 10]
```diff
@@ -28,12 +41,17 @@ def __str__(self) -> str:
...
-            "type": self._derivative_type.value,
+            "contract": self._contract,
+            "underlying": self._underlying,
+            "type": self._derivative_type,
+            "longshort": self._derivative_longshort,
             "strike": self._strike,
             "expiry": self._expiry
         }
 
+    def display(self) -> list[any]:
```
[the remark this answers is not in the document]
Comment (vargala4) [reply src/contract.py:52]: deleted/modified

File: src/contract.py [hunk 11]
```diff
@@ -63,74 +87,322 @@ def get_timeline(self) -> list[float]:
...
-        super().__init__(und, PutCallFwd.FWD, strk, exp)
+    def __init__(self, und: Stock, longshort: LongShort, strk: float, exp: float) -> None:
+        super().__init__(und, PutCallFwd.FWD, longshort, strk, exp)
+        self._contract = ContractType.FORWARD
 
-    def convert_to_generic(self) -> GenericContract:
-        # todo: needs to be properly initialized after GenericContract is implemented
-        return GenericContract()
+
+    def convert_to_generic(self, contract, und, dtype, longshort, strk, exp) -> GenericContract:
```
[the remark this answers is not in the document]
Comment (vargala4) [reply src/contract.py:95]: done

File: src/contract.py (hunk 3 above)
Comment (vargala4) [reply src/contract.py:291]: deleted

File: src/contract.py (hunk 1 above)
Comment (vargala4) [reply src/contract.py:15]: done

File: src/contract.py (hunk 6 above)
Comment (vargala4) [reply src/contract.py:309]: done

File: src/contract.py (hunk 7 above)
Comment (vargala4) [reply src/contract.py:203]: Let us discuss.

File: src/contract.py (hunk 4 above)
Comment (vargala4) [reply src/contract.py:266]: included a new raise function: _raise_incorrect_derivative_type_generic()

File: src/contract.py (hunk 5 above)
Comment (vargala4) [reply src/contract.py:66]: done

File: src/contract.py [hunk 12]
```diff
@@ -1,21 +1,34 @@
...
         self._underlying: Stock = und
         self._derivative_type: PutCallFwd = dtype
+        self._derivative_longshort: LongShort = longshort
+        self.ls = 1 if longshort == LongShort.LONG else -1
         self._strike: float = strk
         self._expiry: float = exp
+        self._num_mon: int = num_mon    # Asian: nr of averaging points; Barrier: nr of monitoring points
+
+    def get_contract(self) -> ContractType:
+        return self._contract
```
[the remark this answers is not in the document]
Comment (friedgab) [reply src/contract.py:21]: We should still have a definition in the `Contract` class for 2 reasons:
- You are using `_contract` in a class where it's neither defined explicitly nor inherited from a base class.
- Nothing guarantees that a subclass of `Contract` defines it.

Either make it an abstract property or just define it as None and let it be inherited.

**Review by friedgab — APPROVED** [review]

**Pull request merged**